### PR TITLE
Add undefined specializations in backward

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43743 Update aliasing in tensorexpr fuser
* #43742 use types in the IR instead of vmap
* #43636 Add inline call to profiling executor pipeline
* #43635 Use prim::TensorExprGroup interned symbol
* #43634 Update requires grad property
* #43633 Specialize optionals for grad_sum_to_size
* **#43632 Add undefined specializations in backward**
* #43631 refactor tests
* #43630 Refactor pass to class
* #43629 Add API for unexecuted op

Specialize the backward graph by guarding on the undefinedness of the input tensors. The graph will look like:
```
ty1, ty2, succesful_checks = prim::TypeCheck(...)
if (succesful_checks)
-> optimized graph
else:
-> fallback graph
```

Specializing on the undefinedness of tensors allows us to clean up the 
```
if any_defined(inputs):
 outputs = <original_computation>
else:
 outputs = autograd zero tensors
```
blocks that make up the backward graph, so that we can fuse the original_computation nodes together.

Differential Revision: [D23358808](https://our.internmc.facebook.com/intern/diff/D23358808)

### CODE REVIEW

File: torch/csrc/jit/passes/specialize_autogradzero.cpp [hunk 1]
```diff
@@ -10,15 +10,49 @@ namespace jit {
...
       : graph_(std::move(graph)) {}
 
   void run() {
-    setStatesOnGraphInputs();
-    specializeAutogradOps(graph_->block());
+    if (!isBackwardGraph()) {
+      return;
+    }
+    if (getProfilingMode()) {
+      if (auto vif = guardSpecializations()) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/specialize_autogradzero.cpp:22]: Does `vif` stand for a "versioning if"? A comment could be helpful here I think.

File: torch/csrc/jit/passes/specialize_autogradzero.cpp [hunk 2]
```diff
@@ -42,6 +76,98 @@ struct AutogradZeroSpecializer {
...
+      }
+      checks.push_back(check);
+    }
+
+    // unable to specialize any of the inputs
+    if (checks.size() == 0) {
+      vif->destroy();
+      // the checks we inserted will be cleaned up
+      // by any subsequent DCE pass
+      return nullptr;
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/specialize_autogradzero.cpp:133]: Nit: might be helpful to have a `GRAPH_DEBUG` statement here.

File: torch/csrc/jit/passes/specialize_autogradzero.cpp [hunk 3]
```diff
@@ -10,15 +10,49 @@ namespace jit {
...
+            default:
+              return false;
+          }
+        });
+  }
+
+  void replaceBlockInputsWithGraph(Block* b) {
+    TORCH_INTERNAL_ASSERT(graph_->inputs().size() == b->inputs().size());
+    for (int i = graph_->inputs().size() - 1; i >= 0; i--) {
+      b->inputs()[i]->replaceAllUsesWith(graph_->inputs()[i]);
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/specialize_autogradzero.cpp:51]: Nit: maybe be do it in two loops to make clang-tidy happy.

**Review by ZolotukhinM — APPROVED** [review]
Looks good!

File: torch/csrc/jit/passes/specialize_autogradzero.cpp [hunk 4]
```diff
@@ -42,6 +76,98 @@ struct AutogradZeroSpecializer {
...
+      vif->destroy();
+      // the checks we inserted will be cleaned up
+      // by any subsequent DCE pass
+      return nullptr;
+    }
+
+    Value* bool_list =
+        graph_->insertNode(graph_->createList(BoolType::get(), checks))
+            ->output();
+    Value* conjunction = graph_->insert(aten::all, {bool_list});
```
Comment (Krovatkin) [inline torch/csrc/jit/passes/specialize_autogradzero.cpp:139]: this is so much  more beautiful then a bunch of `ands`!!!

**Review by Krovatkin — APPROVED** [review]
:shipit:

**Pull request closed without merging**